ASoC: SOF: ipc4-pcm: support multiple configs in BE DAIs

kv2019i opened the pull request:
Depends on: https://github.com/thesofproject/sof/pull/7091

ASoC: SOF: ipc4-pcm: support multiple configs for BE DAIs

Backend DAIs may support multiple audio formats. Modify pipeline
setup to select a suitable configuration based on topology and
frontend DAI runtime configuration.

For sampling rate, if one of the BE DAI configurations has
a sampling rate matching that of FE DAI, configure BE DAI to
this rate.

For sample format, the current code hardcodes DAI copier sample format
to 32bit for both playback and capture pipelines. This is not always
desired, so lift the limitation and set the sample format based on
topology definitions for the copiers. For capture pipelines, we want to
set the BE DAI pipeline format based on topology instead of using the FE
DAI format. This covers the common use-case where BE DAI outputs data at
a higher sample precision and sample width is reduced later in the
pipeline. Instead of hardcoding to 32bit, use the BE DAI copier output
format defined in topology.

kv2019i commented:
@juimonen FYI, this is needed to have the hwconfig selection work with IPC4.

kv2019i commented:
V2 (Jan 24th) update:
- updated implementation and tested with all Bluetooth offload configurations
- cleaned up the code so it is ready for review

juimonen reviewed (dismissed):
for me it looks good, although can't predict if it has other consequences, driver experts can chime in on that.

kv2019i commented:
Hmm, test results are brutal https://sof-ci.01.org/linuxpr/PR4137/build2829/devicetest/index.html so this will need more work. Back to draft and tagging with DNM.

kv2019i commented:
Update on Jan 25th, but results are still bad -> https://sof-ci.01.org/linuxpr/PR4137/build2847/devicetest/index.html 

There's a problem in deciding which copier instance does the format conversions if FE and BE formats don't match. Now in CI, the failed cases have S16_LE in FE and a topology where DAI/BE copier only supports S32_LE.

In current code, the format of DAI copier is just overridden to S32.

Not sure yet whether to fix in topology, or in code, and how.

kv2019i commented:
V4 (25th Jan) update:
- a potential fix to format negotiation for capture pipelines: instead of hardcoding to S32, limit pipeline parameters to format of DAI copier

kv2019i commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -457,17 +457,41 @@ static void ipc4_ssp_dai_config_pcm_params_match(struct snd_sof_dev *sdev, const
+	switch (bit_depth) {
+	case 0:
+		/* bit depth not set, ignore */
+		break;
```
With current topologies, this is needed. Not sure whether to keep this in. I'll add a link once I have the tplg PR.

ranj063 replied:
why would we not set the bit_depth? this seems like a miss

kv2019i replied:
@ranj063 https://github.com/thesofproject/sof/pull/6994

Large number of tests will fail without this. If #6994 was a one-off error, I can remove the "case 0" and make it a fatal error.

kv2019i commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1236,6 +1236,25 @@ static int snd_sof_get_nhlt_endpoint_data(struct snd_sof_dev *sdev, struct snd_s
 }
 #endif
 
+static int ipc4_set_fmt_mask(struct snd_mask *fmt, unsigned int bit_depth)
```
This helper is duplicate here as ipc4-pcm.c needs special treatment for 'bit_depth of zero' case.

kv2019i reviewed (commented):
A few comments to help review.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -495,17 +519,51 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+			fmt = &available_fmt->out_audio_fmt[i];
+			bit_depth = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(fmt->fmt_cfg);
+			if (ipc4_set_fmt_mask(fmt_mask, fmt->bit_depth) < 0)
+				return -EINVAL;
```
@kv2019i I am having a hard time understanding the logic with the loop here. For every available format, you first set the format mask based on the input format and then override it with the output format. Can you please add some explanation as to what we're trying to do here?

kv2019i replied:
@ranj063 The "set" verb is misleading but I'm sticking to ALSA semantics where "snd_mask_set()" actually does a logical OR to the mask, i.e. formats are incrementally added to the mask.

I do agree this is still confusing. This code is in dai_link_fixup, and the dailink is really bidirectional. What we want here is to limit the DAI link to the formats declared in the topology. Unfortunately it is possible to have different formats as copier inputs and outputs. Here I'm declaring all formats (both at input and outputs). This defines a superset (e.g. both S16 and S32), even if for particular direction (like capture) only a a subset is actually supported (S32).

This still seems improvement over existing code where we just hardcode to S32 with no input from topology.

But, but, one option I'm considering whether it would be better not to limit the formats in dai_link_fixup at all. Having the format negotiation done in the path enabling code (e.g. in copier_prepare), might be sufficient and having this in one place is probably easier to understand.
UPDATE: I tried a version that removes modifying fmt mask at all, and seems to work for both old and new usages. So this is an option as well. 

In IPC3 code we do look up more constraints from tplg (e.g. format from private->comp_dai->config.frame_fmt) and set them in dai_link_fixup. And thus I tried to do the same for IPC4.

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1341,16 +1360,24 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 			/*
-			 * modify the input params for the dai copier as it only supports
-			 * 32-bit always
+			 * As path walk starts from the capture dai, and copiers can
```
@kv2019i this is also very hard to follow. Basically, the reason for limiting the input format to 32 bits in the case of the DAI copier was that we can match and retreive the format based on the output format along from the list of available formats. Now, if you want to change the logic to not limit to 32 bits, then what you need is to change the matching function (which is sof_ipc4_init_audio_fmt) to include 2 references not one. The current function will match only the output format based on pipeline_params. But in your case you need to match the output to match pipeline_params and input to also match the runtime params.

kv2019i replied:
@ranj063 I did consider extending sof_ipc4_init_audio_fmt(), but that's used by all modules and this special handling really applies to "edge" copier. In short, when we start the traversal for capture, prepare is called with hw_params initialized to FE. Now the problem is, if the BE supports multiple formats, which should be used. Example:

- arecord opens FE with S16
tplg defines capture pipeline with multiple processing components - configured with S32 input/output
- the BE supports both S16 and S32

With current code, the above works as BE capabilities are ignored and the format is just hardcoded to S32 at the BE DAI (for capture). 

The above does not work if we want to have anything else than S32 flowing from a capture BE DAI. And e.g. for Bluetooth offload we specifically want this (and probably for many other cases as well).

So then we reach the problem:
   which format to pick in copier_prepare for the capture DAI walk?

I think we really want to push this decision to tplg definition and not hardcode it in kernel. Given it is very difficult to try out all combinations (as you'd need to do the full graph walk to see whether all modules on the path can be configured e.g. if copier starts to output S24), what my patch does is to pick the first format defined for the DAI copier as output format.

This allows tplg creators to defined how the data flows out from copiers in capture pipelines and make the format selection more predictable.

So that's the thought process. If this seems feasible, I can try to map that to better code comments.

FYI to @ujfalusi and @plbossart as well.

ranj063 replied:
Hi @kv2019i I'm a bit confused as to how this works. So you want to allow 16-bit capture in the DAI but here you dont change the ref_audio_fmt or the ref_params? Shouldn't you set the ref_params to fe_params and the ref_audio_fmt to the input_params?

kv2019i replied:
@ranj063 More specifically, I'd want to give control of how DAI copier is configured here to topology. The copier can do conversions and in current code, kernel hardcodes this to S32 irrespectively of what is defined in topology for this copier. With this patch, we take the DAI copier output format, and limit the pipeline parameters to its output format.

One confusion is that "fmt" is a mask pointing to pipeline_params and on line 1386 this is copied to ref_params. So we do modify the ref_params which is passed to sof_ipc4_init_audio_fmt().

In sof_ipc4_init_audio_fmt() we have a bit of redundancy as the format selection is already limited, but this is a) the case already as current code overrides fmt to S32, and b) sof_ipc4_init_audio_fmt() is used for all modules so it needs to be more generic.

kv2019i commented:
Test results much better now, marking ready for review -> https://sof-ci.01.org/linuxpr/PR4137/build2853/devicetest/index.html

bardliao commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -495,17 +519,51 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+
+			bit_depth = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(fmt->fmt_cfg);
+			if (ipc4_set_fmt_mask(fmt_mask, bit_depth) < 0)
+				return -EINVAL;
```
Should we return the return value of `ipc4_set_fmt_mask`? Although it is -EINVAL, too.

kv2019i replied:
Ack, fixed (this code removed based on the other discussion on need to modify format mask in dai_fixup).

kv2019i commented:
@RanderWang Can you also take a look?

kv2019i commented:
V5 (Jan 30th):
- removed the format mask modification in dai_fixup (this is not strictly needed and based on feedback from review, code was hard to follow)

RanderWang reviewed (dismissed):
good solution!

kv2019i commented:
V6 (2nd Feb):
- Addressed compiler warning picked up by CI (unsused 'fmt' variable).
- Rewrote the commit description and inline code comments based on review discussion in this PR. The solution is limited in scope, so I expanded the descriptions to explain the logic that is used for sampling rate and sample format, and why the two get slightly different treatment. Hopefully this helps also to check alignment with #4163

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -495,17 +496,32 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+		struct sof_ipc4_available_audio_format *available_fmt = &ipc4_copier->available_fmt;
+		int i;
+
+		for (i = 0; i < available_fmt->audio_fmt_num; i++)
```
@kv2019i do you need to rebase your PR to take in my changes? We no longer have audio_fmt_num

kv2019i replied:
Ack, rebased!

kv2019i commented:
V7 (9th Feb):
- rebased on top of latest changes how input/output pin audio formats are handled
- retested with nocodec BT test topology

kv2019i commented:
@ranj063 Rebased now. Although the second patch to unify common code might need update/refresh for 
https://github.com/thesofproject/linux/pull/4185 which one more instance of code to go from IPC4 sample format into ALSA format enum. Small thing though, but we'll have to anyways have a clean set for upstreaming...

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1339,16 +1358,29 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 			fmt = hw_param_mask(pipeline_params, SNDRV_PCM_HW_PARAM_FORMAT);
 			snd_mask_none(fmt);
-			snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+			fmt_cfg = available_fmt->output_pin_fmts[0].audio_fmt.fmt_cfg;
```
@kv2019i this change to use the output format in topology is only because we only have 32-bit support in the topology today right? In theory we should just be able to use the fe_params for capture and pipeline_params for playback right?

kv2019i replied:
@ranj063 We don't want the pipeline to be 16bit just because FE is 16bit. This should be configured in topology (e.g. if you want higher precision processing in the processing pipeline). I.e. topology author can choose where precision is reduced in the pipeline.

ranj063 replied:
@kv2019i if the pipeline requires a different format for processing, the copier will do the conversion no? Of course,.this will avoid the conversion.

kv2019i replied:
@ranj063 Right but we are configuring the first copier here and as many components can support multiple formats, if we set S16 here, the pipeline will then be S16. There are multiple possible ways to configure the pipeline in these cases, but I think both a) configure always to S32 (current model), or b) locking to FE, are too simple policies.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -495,17 +496,36 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+		rate->min = fe_rate;
+		rate->max = rate->min;
+	} else {
+		rate->min = ipc4_copier->available_fmt.input_pin_fmts->audio_fmt.sampling_frequency;
```
@kv2019i the if part is clear, the else not so much. If there's no matching audio format with the fe sampling rate, then why would we choose the first available input arbitrarily? Is this assuming that in this case there will be an SRC in the pipeline that will do the sample rate conversion and that all input formats in topology have the same sampling rate?

kv2019i replied:
@ranj063 That's what the existing code does? :) As far as I understand the existing logic (and it's similar in e.g. ipc3-pcm) that we want to limit the sampling rate to that defined in topology. So in vast majority of cases, there's just one rate defined and that is used to set the constraint.

If we have multiple rates in tplg, then we need to pick one and using the FE rate seems to be a good policy here.

If there's no match, then there's either a SRC in the pipeline or the pipeline setup should rightfully fail. We don't really have knowledge in this dai fixup, which rate should we constrain (either multiple options or one option that does not match FE). I remember trying to just not limit the rate here if no match is found with FE, but resulted in many failures. I do think we need to limit the rate somehow here based on topology information.

kv2019i replied:
Seems the else-path has limited function as we are checking the formats when we enable the path, so doing the limitation in dai_fixup is maybe not needed.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -495,17 +496,36 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+				&available_fmt->output_pin_fmts[i].audio_fmt;
+
+			if (fe_rate == audio_fmt->sampling_frequency)
+				fe_be_rate_match = true;
```
break out from the loop here?

kv2019i replied:
right, I thought I had some need to go through the full loop, but alas, it's not needed. will fix in V9, thanks

ujfalusi replied:
you can still break out from here...
the `single_be_rate` does not matter if `fe_be_rate_match` is true.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -495,17 +496,36 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+	if (fe_be_rate_match || use_chain_dma) {
+		/* Limit to common FE/BE sampling rate */
+		rate->min = fe_rate;
+		rate->max = rate->min;
```
Do we need to touch these at all?
`rate->min` _is_ `params_rate(params);`
Not sure about the rate->mask, but since this is hw_params callback, In think rate->min == rate->max

kv2019i replied:
@ujfalusi You are probably right and this could be left out. For V9.

ujfalusi replied:
I still think that this is a NOP.

kv2019i commented:
V8 (10th Feb):
- add explicit handling in case there are multiple formats defined both when selecting sampling rate and the format
- if topology defines multiple formats for DAI copier output, but none of them match FE, raise an error as kernel cannot decide what is the best solution 

This is getting increasingly complex to handle various possibilities of what could be configured in tplg. But at least this V8 version does not have any logic that assumes anything based on "first input/output format".

UPDATE: I expected failures in CI as we have existing topologies that define multiple output formats for DAI copiers and with PR, kernel will raise an error (not a valid topology definition)

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -463,11 +463,14 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 	bool use_chain_dma = false;
+	unsigned int fe_rate;
+	bool fe_be_rate_match = false;
+	bool single_be_rate = true;
```
if these were be in a nice reverse tree style?

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1222,6 +1222,68 @@ static int snd_sof_get_nhlt_endpoint_data(struct snd_sof_dev *sdev, struct snd_s
+
+	/* case 3: single format defined in topology, use that */
+	return ipc4_set_fmt_mask(fmt, sample_valid_bits);
+}
```
case 4: single format defined in topology and it is not matching?

kv2019i replied:
@ujfalusi that's case 3, if there's just a single format, we use that whether it matches with FE or not. the pipeline is guaranteed to have another copier that can do the conversion, so this should be always good. topology developer wanted to use this format, so let's use what is requested.

case-1 is a bit different as kernel needs to pick one, but has no context to make a good choice. should we pick the first one, or the format with highest precision...?

kv2019i commented:
So indeed V8 hit issues:
https://sof-ci.01.org/linuxpr/PR4137/build3171/devicetest/index.html?model=ADLP_RVP_SDW_IPC4ZPH&testcase=check-capture-10sec

This comes down to "if topology defines multiple formats for DAI copier output, but none of them match FE, raise an error as kernel cannot decide what is the best solution" -> feedback @ranj063 @ujfalusi @aiChaoSONG on this, should we change tplgs to conform to this, or should I use a more relaxed policy in this PR?

kv2019i commented:
V9 (13th Feb, commit 42241c32874):
- emit a warning if topoplogy has multiple formats for dai_out (instead of raising an error) -> should fix the CI failures as there were many such cases in our current mainline tplgs
- fixed based on feedback from @ujfalusi to V8

kv2019i commented:
V10 (Feb 13th, commit ca796dbf8de):
- one more issue raised by @ujfalusi fixed
- regression in V9 for chain-dma

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1222,6 +1222,70 @@ static int snd_sof_get_nhlt_endpoint_data(struct snd_sof_dev *sdev, struct snd_s
+	snd_mask_none(fmt);
+
+	if (multiple_formats) {
+		if (fe_format_match) {
```
@kv2019i if multiple_formats is true and fe_format_match is untrue, we should return an error instead of returning an arbitrary format with a warning

kv2019i replied:
@ranj063  Ack. We'll need to rework the tplgs before we can do that in kernel. I started this work now in https://github.com/thesofproject/sof/pull/7091

ujfalusi replied:
`multiple_formats == false` && `fe_format_match == false` is a possibility? The we will set the first bit_depth, right?

kv2019i commented:
V11 (Feb 14th, commit 0d0051652a53)
- reverted to stricter policy on tplg definitions for dai_out
- above change will result in CI failures with current SOF main, at least https://github.com/thesofproject/sof/pull/7091 will be needed (possibly more PRs to SOF)

ujfalusi commented:
> V11 (Feb 14th, commit [0d00516](https://github.com/thesofproject/linux/commit/0d0051652a53e8c362c614b1edabd37b46b40918))
> 
>     * reverted to stricter policy on tplg definitions for dai_out
> 
>     * above change will result in CI failures with current SOF main, at least [202302 tplg add capture be pipe sof#7091](https://github.com/thesofproject/sof/pull/7091) will be needed (possibly more PRs to SOF)

and https://github.com/thesofproject/sof/pull/7091 needs this PR to pass CI or the dependency is not circular?

kv2019i commented:
@ujfalusi wrote:
> and [thesofproject/sof#7091](https://github.com/thesofproject/sof/pull/7091) needs this PR to pass CI or the dependency is not circular?

One way only, the FW 7091 should work with current kernel. But once kernel starts being stricter about the tplg definitions, then FW needs to be ready.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -495,17 +496,44 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+		if (!fe_be_rate_match && single_be_rate) {
+			/* Only one format defined, use that */
+			rate->min = available_fmt->input_pin_fmts[0].audio_fmt.sampling_frequency;
+			rate->max = rate->min;
```
It took me some time to get this, but the comment does not helped...
And I'm still not sure if I got it...

`available_fmt->input_pin_fmts[0].audio_fmt.sampling_frequency` is `be_rate`, right?

no, I don't get what this is doing.

What happens if the `available_fmt->input_pin_fmts[]` only contains single rates, but different sample sizes, channels and this rate does not match for the fe_rate?
We will fail somewhere else?

kv2019i replied:
@ujfalusi Right, now that I moved (in V11) this piece of code to same scope where 'be_rate'  is defined, yes, I could assign to 'be_rate' now. Will fix the comment.

For the bigger question of what happens if rate is not limited at all, that's a good question. I've been just going by what the existing code does and what we've done with ipc3-pcm.c in the past. For the specific need of enabling multiple BE configs, I've not wanted to reopen the wider arch.

But there is certainly overlap. The dai-fixup is called in hw_params flow. This is standard ASoC stuff. Then in sof_pcm_hw_params() we run the sof_pcm_setup_connected_widgets() path which will do a more complex path enablement and this logic is more specific to SOF. E.g. for capture we traverse all widgets on the path, and set formats for each.

Now whether all constraints could be moved to sof_pcm_setup_connected_widgets(), I don't really know. I don't really see a big obstacle really. @ranj063 @plbossart ? Both paths of logic are done at hw_params epoch, so I don't see a big difference, but I'm sure there is some subtle dependencies.

kv2019i replied:
@ujfalusi Dug the history of this code a bit more and the key reason to have the "rate->min/max" lookup in dai fixup is that we do the SSP blob selection here as well (via ipc4_ssp_dai_config_pcm_params_match()). So that's one option to hide this logic behind the SSP specific code and only limit the rate if we need to choose a blob. At least this would be hidden deeper in vendor/DAI specific code and not have this in generic fixup code for IPC4....

RanderWang replied:
@kv2019i Another reason is for SRC, .e.g. 44.1k stream for FE, we do SRC in fw to convert it to 48k supported by BE, so we need to use this fixup to change rate for audio bus controller and codec driver.

RanderWang replied:
Maybe it is caused my first patch, now I prefer to use output format available_fmt->output_pin_fmts[0].audio_fmt.sampling_frequency since output is for audio bus and codec. we may have different input valid bit from internal pipeline but only one output valid bits by audio dsp

kv2019i replied:
@RanderWang fixup is called for both directions (capture/playback) so use of copier pins configs is problematic as e.g. for playback case, copier input may be different than what is actually sent to the DAI. Especially to evaluate sample formats (as copier can do conversions and input/output formats may be different), this is a real problem. For sampling rate (as used in this PR) and channel count, it should not matter whether we look at input or output pins, as both must be the same (as copier cannot convert). This is harder to follow in general than IPC3, where we had an explicit object to describe DAI properties.

RanderWang commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -495,17 +496,44 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+	 * Limit sampling rate if copier has a format that matches
+	 * the FE rate.
 	 */
 	if (!use_chain_dma) {
```
so now if (use_chain_dma) we do nothing but the original code set rate->min and rate->max.  If we don't need to set rate min and max, we can do it as
````c
if (use_chain_dma)
     return 0
````

ujfalusi replied:
We have a switch for dai_type later on, yes it is only handling INTEL_SSP and the ChainDMA is only supported (for now) via HDA, but that can change, I think. Probably  a goto then?

kv2019i replied:
@RanderWang I was wondering about this but in theory we could use chain-dma for SSP as well, and in that case, we may need to run the DAI-type specific logic later in the function. But yeah, this code could be more reabable, let me split this segment to a separate subfunction in next version.

ranj063 replied:
We cannot use chain DMA for SSP. It uses gpdma

ujfalusi replied:
@ranj063, #4158 gives a glimpse into the future ;)
But the ChainDMA FW implementation as it is does not rule out use of GPDMA on the peripheral side, with some effort it can be made working.

ujfalusi replied:
Having said that, for now a simple return 0 in case of chain_dma would make the code nicer. If there will be a need for a change, it can be done when it is due.

kv2019i replied:
@ranj063 @ujfalusi As it's more a FW side development whether chain_dma can work with SSP, I'd rather keep the kernel code neutral on this.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -495,18 +547,14 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
-		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+		ret = sof_ipc4_pcm_dai_link_fixup_rate(sdev, params, ipc4_copier);
+		if (ret < 0)
+			return ret;
```
nitpick: you only need the `ret` within this curly block.

kv2019i commented:
V12 (Feb 15th, commit 2af0d033fbd1):
- reworked the dai_fixup code and code comments for better readability (based on feedback from @ujfalusi  and @RanderWang )
- raise an error if copier config in tplg does not provide sufficient information to pick up correct sampling rate (feedback from @ujfalusi )
- the FW PR updated https://github.com/thesofproject/sof/pull/7091 and now passes the PR test plan together with this PR (FW PR needs to be merged first)

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -457,18 +457,70 @@ static void ipc4_ssp_dai_config_pcm_params_match(struct snd_sof_dev *sdev, const
+		&ipc4_copier->available_fmt;
+	struct sof_ipc4_audio_format *audio_fmt =
+		&available_fmt->input_pin_fmts[0].audio_fmt;
+	unsigned int be_rate  = audio_fmt->sampling_frequency;
```
this is one of those rare cases where I would _not_ initialize at declaration to make it easier to see what we are trying to do.
Actually, this is not that ugly (85 char long line):
```c
u32 be_rate = available_fmt->input_pin_fmts[0].audio_fmt.sampling_frequency;
struct sof_ipc4_audio_format *audio_fmt;
```
sampling_frequency is `uint32_t` after all.

kv2019i replied:
@ujfalusi I'll keep "unsigned int" as "params_rate(params)" returns unsigned int (sound/pcm.h). Mixing u32 and unsigned int seems a bit sloppy for little gain.

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -457,18 +457,70 @@ static void ipc4_ssp_dai_config_pcm_params_match(struct snd_sof_dev *sdev, const
+	bool single_be_rate = true;
+	int i;
+
+	/*
```
Not  sure if it helps or make any sense here, but:
```c
/* No need to check further if we have a match to begin with */
if (fe_rate == be_rate)
	return 0;
```

kv2019i replied:
@ujfalusi Wouldn't this be a bit misleading as we may have multiple be_rates? Sure this is correct, but the loop will get you the same result. I can add if this seems clearer of course.

ujfalusi replied:
sure, the first iteration will test in the loop `if (be_rate == fe_rate)` and if this is true, then we break out and `if (!fe_be_rate_match)` would not be taken so we return with 0.

kv2019i replied:
@ujfalusi Didn't add the additional check now in V13.

ujfalusi replied:
OK

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1222,6 +1222,69 @@ static int snd_sof_get_nhlt_endpoint_data(struct snd_sof_dev *sdev, struct snd_s
+
+		if (snd_pcm_format_width(params_format(fe_params)) == val)
+			fe_format_match = true;
+	}
```
I think if I look at this loop and the one in `sof_ipc4_pcm_dai_link_fixup_rate()` they are doing exactly the same thing, one with rate the other with valid_bits, yet the implementation of them are unique.
I would consider using  the same logical flow in both to make things simpler.

ujfalusi replied:
Right, the verbs are negated in between the two, but they do the same, if I'm not mistaken?

kv2019i replied:
@ujfalusi It's similar but not the same. The rate check is symmetrical and can be done for playback and capture, but the format check is only makes sense for capture (dai_out) due to capabilities of the copiers. I could align the code to look more the same indeed.. makes sense.

ujfalusi replied:
Yes, I got that part, what I'm saying that the two loops can be made identical looking, in essence the only diff is _what_ is compared, but the flow can be the same

```c
	for (i = 0; i < available_fmt->num_output_formats; i++) {
		unsigned int val;

		audio_fmt = &available_fmt->output_pin_fmts[i].audio_fmt;
		val = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(audio_fmt->fmt_cfg);

		if (i == 0)
			sample_valid_bits = val;
		else if (sample_valid_bits != val)
			multiple_formats = true;

		if (snd_pcm_format_width(params_format(fe_params)) == val)
			fe_format_match = true;
	}
```
VS (this is what you have in `sof_ipc4_pcm_dai_link_fixup_rate()`)
```c
	audio_fmt = &available_fmt->output_pin_fmts[i].audio_fmt;
	sample_valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(audio_fmt->fmt_cfg);

	for (i = 0; i < available_fmt->num_output_formats; i++) {
		unsigned int val;

		audio_fmt = &available_fmt->output_pin_fmts[i].audio_fmt;
		val = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(audio_fmt->fmt_cfg);

		if (snd_pcm_format_width(params_format(fe_params)) == val)
			fe_format_match = true;

		if (sample_valid_bits != val)
			single_formats = false;
	}
```

right, this is nitpicking.

kv2019i replied:
@ujfalusi please check, tried to use same style in both functions now.

ujfalusi replied:
Thanks, looks confusingly similar :D

ranj063 reviewed (dismissed):
Looks good @kv2019i

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -495,18 +547,14 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
-		snd_mask_none(fmt);
-		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+		ret = sof_ipc4_pcm_dai_link_fixup_rate(sdev, params, ipc4_copier);
+		if (ret < 0)
```
nitpick: `if (ret)`

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1493,21 +1553,7 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
-			params_format(pipeline_params));
-		return -EINVAL;
-	}
+	ret = ipc4_set_fmt_mask(fmt, out_sample_valid_bits);
```
probably check the ret?

kv2019i commented:
V13 (Feb 16th, commit 599f71dcd626):
- harmonized the code style in sof_ipc4_pcm_dai_link_fixup_rate() and ipc4_copier_set_capture_fmt()
- other fixes to comments raised in V12 review 
- the FW dependency is now merged, so expectation is that CI will no pass for this PR

ujfalusi reviewed (approved):
@kv2019i, from code point of view looks nice and polished, thank you!
Let's see what CI thinks now.

ujfalusi commented:
@kv2019i, well, it is not even compiling :o

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -457,18 +457,70 @@ static void ipc4_ssp_dai_config_pcm_params_match(struct snd_sof_dev *sdev, const
 	struct sof_ipc4_copier *ipc4_copier;
 	bool use_chain_dma = false;
-	int dir;
+	int dir, ret;
```
this `ret` is no longer needed

kv2019i commented:
V14 (Feb 16th, commit f37e1d237109)
- fix compiler warning (raised as error in CI build), no other changes

plbossart reviewed (approved):
I am not sure what happens if we have multi stages of mixers/demux and changes of frequency and formats on intermediate branches, but for now this looks good enough. Nice work @kv2019i
